collectors/recon.go

```diff
@@ -0,0 +1,637 @@
+		errCount++
+		sendReconErrorCount(ch, cmdArgs, errCount)
+		return
+	}
+
+	gotErr := false
+	for hostname, dataBytes := range result {
+		var disksData []struct {
+			Device  string `json:"device"`
+			Avail   int64  `json:"avail"`
+			Mounted bool   `json:"mounted"`
+			Used    int64  `json:"used"`
+			Size    int64  `json:"size"`
+		}
+		err = json.Unmarshal(dataBytes, &disksData)
+		if err != nil {
+			logg.Error("recon collector: %s: %s: %v", taskName, hostname, err)
+			gotErr = true
+			continue
+		}
+
+		var totalFree, totalUsed, totalSize int64
+		for _, disk := range disksData {
+			if !(disk.Mounted) {
+				continue
+			}
+
+			totalFree += disk.Avail
+			totalUsed += disk.Used
+			totalSize += disk.Size
+
+			// submit metrics by disk (only used percent here, which is the
+			// most useful for alerting)
+			device := specialCharRx.ReplaceAllLiteralString(disk.Device, "")
+			ch <- prometheus.MustNewConstMetric(
+				clusterStorageUsedPercentByDiskDesc,
+				prometheus.GaugeValue, float64(disk.Used)/float64(disk.Size),
+				hostname, device,
+			)
+		}
+
+		usedPercent := float64(totalUsed) / float64(totalSize)
+		if totalSize == 0 {
+			usedPercent = 1.0
+		}
+		ch <- prometheus.MustNewConstMetric(
+			clusterStorageUsedPercentDesc,
+			prometheus.GaugeValue, usedPercent,
+			hostname,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			clusterStorageUsedBytesDesc,
+			prometheus.GaugeValue, float64(totalUsed),
+			hostname,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			clusterStorageFreeBytesDesc,
+			prometheus.GaugeValue, float64(totalFree),
+			hostname,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			clusterStorageCapacityBytesDesc,
+			prometheus.GaugeValue, float64(totalSize),
+			hostname,
+		)
+	}
+	if gotErr {
+		errCount++
+		sendReconErrorCount(ch, cmdArgs, errCount)
+	}
+}
+
+var (
+	clusterMD5AllDesc = prometheus.NewDesc(
+		"swift_cluster_md5_all",
+		"Sum of matched-, not matched hosts, and errors encountered while check hosts for md5sum(s) as reported by the swift-recon tool.",
+		[]string{"kind"}, nil,
+	)
+	clusterMD5MatchedDesc = prometheus.NewDesc(
+		"swift_cluster_md5_matched",
+		"Matched hosts for md5sum(s) reported by the swift-recon tool.",
+		[]string{"kind"}, nil,
+	)
+	clusterMD5NotMatchedDesc = prometheus.NewDesc(
+		"swift_cluster_md5_not_matched",
+		"Not matched hosts for md5sum(s) reported by the swift-recon tool.",
+		[]string{"kind"}, nil,
+	)
+	clusterMD5ErrorsDesc = prometheus.NewDesc(
+		"swift_cluster_md5_errors",
+		"Errors encountered while checking hosts for md5sum(s) as reported by the swift-recon tool.",
+		[]string{"kind"}, nil,
+	)
+)
+
+var reconMD5Rx = regexp.MustCompile(
+	`(?m)^.* Checking ([\.a-zA-Z0-9_]+) md5sum(?:s)?\s*([0-9]+)/([0-9]+) hosts matched, ([0-9]+) error.*$`)
+
+func reconMD5Task(taskName, pathToReconExecutable string, ch chan<- prometheus.Metric) {
+	errCount := 0
+	cmdArgs := []string{"--md5"}
+	sendReconErrorCount(ch, cmdArgs, errCount)
```

Comment:
When error paths are taken down below, `sendReconErrorCount()` will be called multiple times, which will cause multiple metrics with the same set of labels to be emitted. Prometheus probably won't like that. I suggest you change it like this:

```suggestion
	defer sendReconErrorCount(ch, cmdArgs, errCount)
```

and remove all the other `sendReconErrorCount()` invocations in this function. This causes the error count metric to be submitted when the function returns, using the then-current value of `errCount`.

(Same for all the other task functions.)